File: practice/programmers/베스트앨범.py
def solution(genres, plays):
    d = {}
    answer = []


    for i in range(len(genres)):
        if d.get(genres[i]) == None:
            songs = {}
            songs[i] = plays[i]
            d[genres[i]] = [plays[i],songs]# [sum, dictionary of song #]
        else:
            d[genres[i]][0] += plays[i]
            d[genres[i]][1][i] = plays[i]
    print(d.items())
    genre_sort = sorted(d.items(), key=lambda x:x[1][0], reverse= True)
    for each in genre_sort:
        song_dict = each[1][1]
        play_sort = sorted(song_dict.items(), key=lambda x:x[1], reverse= True)
        answer.append(play_sort[0][0])
        if len(play_sort) > 1:
            answer.append(play_sort[1][0])
    print(answer)
    return answer

solution(["classic", "pop", "classic", "classic", "pop"], [500, 600, 150, 800, 2500]
)

File: practice/programmers/js/베스트앨범.js
function compareFunction(
  [aIndex, aPlays, _, aTotalPlays],
  [bIndex, bPlays, __, bTotalPlays]
) {
  if (aTotalPlays > bTotalPlays) {
    return -1;
  }
  if (aTotalPlays < bTotalPlays) {
    return 1;
  }
  if (aPlays > bPlays) {
    return -1;
  }
  if (aPlays < bPlays) {
    return 1;
  }
  if (aIndex < bIndex) {
    return -1;
  }
  if (aIndex > bIndex) {
    return 1;
  }
  return 0;
}

function solution(genres, plays) {
  const albumMap = genres.reduce((acc, curVal, curIndex) => {
    acc[curVal] = acc[curVal]
      ? [...acc[curVal], [curIndex, plays[curIndex], curVal]]
      : [[curIndex, plays[curIndex], curVal]];
    return acc;
  }, {});

  for (const [genre, songs] of Object.entries(albumMap)) {
    const totalPlays = songs.reduce((acc, [_, plays, __]) => {
      return acc + plays;
    }, 0);
    albumMap[genre] = albumMap[genre].map((song) => [...song, totalPlays]);
  }

  let allSongs = [];
  for (const [_, songs] of Object.entries(albumMap)) {
    allSongs = [...allSongs, ...songs.sort(compareFunction).slice(0, 2)];
  }
  const answer = allSongs.sort(compareFunction);
  return answer.map(([index, _, __, ___]) => index);
}

solution(
  ["classic", "pop", "classic", "classic", "pop"],
  [500, 600, 150, 800, 2500]
);
